crates/bevy_ecs/src/observer/builder.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::TypeId;
+
+use super::*;
+
+/// Builder struct for [`Observer`].
+pub struct ObserverBuilder<'w, E = ()> {
+    commands: Commands<'w, 'w>,
+    descriptor: ObserverDescriptor,
+    _marker: PhantomData<E>,
+}
+
+impl<'w, E: 'static> ObserverBuilder<'w, E> {
+    /// Constructs a new [`ObserverBuilder`].
+    pub fn new(commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        let event = commands
+            .components()
+            .get_id(TypeId::of::<E>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<E>(),
+                )
+            });
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Constructs an [`ObserverBuilder`] with a dynamic event id.
+    /// # Safety
+    /// Caller must ensure that the component associated with `id` is accessible as E
+    #[must_use]
+    pub unsafe fn new_with_id(event: ComponentId, commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Adds `NewE` to the list of events listened to by this observer.
+    /// Observers that listen to multiple types of events can no longer access the typed event data.
+    pub fn on_event<NewE: 'static>(&mut self) -> &mut ObserverBuilder<'w, ()> {
```

Comment:
This should document the behavior of adding the same `NewE` multiple times. As should the untyped version.

Reply:
True, I hadn't considered someone trying to do that, with the current implementation it would eventually be a no-op as it registers the same observer under the same component twice which will do nothing, however we can detect the duplicate early if we want to.

Reply:
Should probably check early yeah